### PR TITLE
Support general MOE replacement for MOE models (Transformers 5.0 compatible)

### PR DESCRIPTION
**Pull request by lvliang-intel**
## Description

Design Overview
This feature provides a generic, model-agnostic mechanism to enable quantization of Mixture-of-Experts (MOE) models by leveraging transformers 5.0+'s native MOE integration interface (@use_experts_implementation decorator and [ALL_EXPERTS_FUNCTIONS registry).

Core Problem
MOE models (DeepSeek-V2, Llama4, Qwen-MoE, Mixtral, etc.) use fused 3D parameter tensors for expert weights in transformers 5.0.

gate_up_proj: (num_experts, 2*intermediate_dim, hidden_dim)
down_proj: (num_experts, hidden_dim, intermediate_dim)

These fused tensors cannot be directly quantized as nn.Linear layers, which auto-round expects.

Solution Architecture
Transformers 5.0+ MOE Interface: ALL_EXPERTS_FUNCTIONS registry , "linear_loop" (our custom implementation)
auto_round/modelling/moe_experts_impl.py:
linear_loop_experts_forward()
- Loops over experts using nn.Linear layers
- Supports quantized nn.Linear (NVFP4, INT4, etc.)
_unfuse_experts_weights_inplace()
- Converts 3D Parameters → nn.ModuleList[nn.Linear]
- Handles transposed vs non-transposed layouts
prepare_model_for_moe_quantization()
- Main entry point
- Registers linear_loop, unfuses all experts modules

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring
- [ ] Other (please specify):

## Related Issues

[](https://github.com/intel/auto-round/issues/1345)

Fixes or relates to #

## Checklist Before Submitting

- [x] My code has been tested locally.
- [ ] Documentation has been updated as needed.
- [x] New or updated tests are included where applicable.

### CODE REVIEW

File: auto_round/modeling/fused_moe/replace_modules.py [hunk 1]
```diff
@@ -255,8 +318,18 @@ def apply_replacements(
...
+            # The module might have been replaced earlier in the loop (parent-first replacement).
+            # Skip if the class has changed or it no longer matches replacement criteria.
+            if module.__class__.__name__ != class_name:
+                logger.debug(
+                    f"Skipping replacement for {name}: class changed from {class_name} to {module.__class__.__name__}"
+                )
+                continue
             with dump_memory_usage_ctx(f"Replacing module {name}", log_level="debug"):
                 replacement_cls = ReplacementModuleBase.get_replacement_class(class_name)
+                if not replacement_cls.is_to_be_replaced(module):
```
[the remark this answers is not in the document]
Comment (lvliang-intel) [reply auto_round/modeling/fused_moe/replace_modules.py:357]: The get_submodule call is not "unnecessary" - it's essential for correctness when dealing with cascading module replacements.

Correct order:
Re-fetch the module from model (ensure we have the latest)
Check if it has already been replaced
Only proceed with replacement if not already replaced

File: auto_round/utils/model.py [hunk 2]
```diff
@@ -1462,28 +1462,41 @@ def set_amax_for_uncalibrated_experts(
...
         if len(amax_values) == 0:
-            return uncalibrated_experts
-        # Flatten all tensors to 1D before concatenation
-        flat_values = [t.reshape(-1) for t in amax_values]
-        all_values = torch.cat(flat_values)
-        set_amax_value = torch.max(all_values)
+            # All experts are uncalibrated - this indicates insufficient calibration data
+            # We cannot proceed without any reference act_max values
+            logger.warning_once(
+                "All expert layers are missing amax values. "
```
Comment (wenhuach21) [inline auto_round/utils/model.py]: @yiliu30 @WeiweiZhang1  we could activate all experts for this scenario

File: auto_round/modeling/fused_moe/replace_modules.py [hunk 3]
```diff
@@ -44,6 +69,33 @@ def _import_required_replacements(model: torch.nn.Module) -> None:
...
+    This is only true when:
+    1. model.config._experts_implementation == "linear_loop" (set by prepare_model_for_moe_quantization)
+    2. The experts' gate_up_proj and down_proj are already nn.ModuleList
+
+    Note: _experts_implementation is only set when the experts class supports
+    @use_experts_implementation decorator, so we don't need to check that again here.
+    """
+    if not hasattr(model, "config"):
+        return False
+    if getattr(model.config, "_experts_implementation", None) != "linear_loop":
```
Comment (wenhuach21) [inline auto_round/modeling/fused_moe/replace_modules.py]: Since we use the linear loop in many places, it’s better to define a variable to represent it. This way, if Transformers prefers a different name, we can update it easily in one place.

File: auto_round/modeling/fused_moe/moe_experts_interface.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+            if has_bias and gate_up_bias is not None:
+                gate_up_linear.bias.data.copy_(gate_up_bias[i])
+                down_linear.bias.data.copy_(down_bias[i])
+
+        gate_up_linears.append(gate_up_linear)
+        down_linears.append(down_linear)
+
+    # Replace the fused parameters with ModuleLists
+    # First, remove the old parameters
+    del module.gate_up_proj
```
Comment (wenhuach21) [inline auto_round/modeling/fused_moe/moe_experts_interface.py]: Is it possible to make the code more robust for future models? Although current MoE modules share the same architecture, I’m concerned that some new models may not follow this pattern. For example, they might include only a gate or only a down projection.

Would it be possible to specify expert names e.g,[gate_up, down], in the config and handle each module list separately?

Comment (wenhuach21) [reply auto_round/modeling/fused_moe/moe_experts_interface.py]: If possible, we could refine it in the future

File: auto_round/modeling/fused_moe/replace_modules.py (hunk 3 above)
Comment (lvliang-intel) [reply auto_round/modeling/fused_moe/replace_modules.py]: OK, let me refine it.

File: auto_round/modeling/fused_moe/moe_experts_interface.py (hunk 4 above)
Comment (lvliang-intel) [reply auto_round/modeling/fused_moe/moe_experts_interface.py]: Good suggestion, I will update the code.

File: auto_round/export/utils.py [hunk 5]
```diff
@@ -68,6 +131,35 @@ def save_model(
         data["torch_dtype"] = str(dtype).split(".")[-1]
         with open(config_path, "w") as file:
             json.dump(data, file, indent=2)
+
+    # Save _experts_implementation to config.json if set (for MOE models with unfused experts)
+    # Check main config and nested configs (e.g., text_config for VLM models)
+    if hasattr(model, "config") and os.path.exists(config_path):
```
Comment (yiliu30) [inline auto_round/export/utils.py]: If we export the model to LLMC format and write the customized `_experts_implementation` into the JSON, will it cause a loading failure since it’s Autoround’s custom logic?

File: auto_round/modeling/fused_moe/moe_experts_interface.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        hidden_states: Input tensor of shape (num_tokens, hidden_dim)
+        top_k_index: Selected expert indices of shape (num_tokens, top_k)
+        top_k_weights: Expert weights of shape (num_tokens, top_k)
+
+    Returns:
+        final_hidden_states: Output tensor of shape (num_tokens, hidden_dim)
+    """
+    global _linear_loop_logged
+    if not _linear_loop_logged:
+        logger.info(f"Using linear_loop experts forward for {self.__class__.__name__}")
```
Comment (yiliu30) [inline auto_round/modeling/fused_moe/moe_experts_interface.py]: nit: How about using `logger.warning_once`?

File: auto_round/modeling/fused_moe/moe_experts_interface.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        _linear_loop_logged = True
+
+    device = hidden_states.device
+    num_top_k = top_k_index.size(-1)
+    num_tokens = hidden_states.size(0)
```
Comment (yiliu30) [inline auto_round/modeling/fused_moe/moe_experts_interface.py:112]: In case some models use the `[bs, seq_len, dim]` input format, we should reshape it to [num_tokens, dim] first.

File: auto_round/modeling/fused_moe/moe_experts_interface.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+                gate_up_linear.bias.data.copy_(gate_up_bias[i])
+                down_linear.bias.data.copy_(down_bias[i])
+
+        gate_up_linears.append(gate_up_linear)
+        down_linears.append(down_linear)
+
+    # Replace the fused parameters with ModuleLists
+    # First, remove the old parameters
+    del module.gate_up_proj
+    del module.down_proj
```
Comment (yiliu30) [inline auto_round/modeling/fused_moe/moe_experts_interface.py]: We might want to test this on a large model to confirm the original weights really get released. Using `module.gate_up_proj.to_empty(device='meta')` should help free them.

File: auto_round/modeling/fused_moe/replace_modules.py [hunk 9]
```diff
@@ -230,11 +283,18 @@ def apply_replacements(
...
+            (transformers 5.0+ pattern). Default is True.
 
     Returns:
         The model with modules replaced.
     """
     _import_required_replacements(model)
+
+    # Auto-detect and handle fused MOE modules if enabled
+    if auto_detect_moe:
+        _handle_moe_modules(model)
```
Comment (yiliu30) [inline auto_round/modeling/fused_moe/replace_modules.py:323]: Is this compatible with monkey‑patched models? My assumption is that patched modules shouldn’t go through this path.

File: auto_round/utils/model.py (hunk 2 above)
Comment (yiliu30) [reply auto_round/utils/model.py]: I’m a little confused here, even with just one token, we should still see top‑k experts selected. 
If every expert ends up uncalibrated, could that indicate a bug?

File: test/test_cuda/models/test_moe_experts_interface.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+    # Verify
+    assert model.config._experts_implementation == "linear_loop"
+    assert len(unfused) == 1
+    assert isinstance(model.layer["experts"].gate_up_proj, nn.ModuleList)
+
+    print("✓ prepare_model_for_moe_quantization works correctly")
+
+
+if __name__ == "__main__":
```
Comment (yiliu30) [inline test/test_cuda/models/test_moe_experts_interface.py]: nit: we could remove them for pytest.

**Review by yiliu30 — APPROVED** [review]
Solid work! 
Left a few nit comments, overall LGTM!
